### DISCUSSION TITLE
Pagination on Curbs endpoints

### DISCUSSION THREAD

**Discussion by schnuerle** [openmobilityfoundation/curb-data-specification#62] [Data Specification]
The query and fetch endpoints for /curbs/zone return [2 data objects](https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/curbs/examples.md#curb-zones-minimum), zones and policies. 

Since these endpoints [can be paginated](https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/general-information.md#pagination), how should pagination be handled? What is the best practice method? Should all curbs data be returned first, then policies data? Or should it be split evenly as much as possible?

**How you handle keeping state when returning data across multiple endpoint pages?**

**Comment by schnuerle** [comment]
See this discussion @zanebclark. Welcome your thoughts.

**Comment by joeb1415** [comment]
Hi folks, long time listener first time commenter :) 

I'd prefer to keep one endpoint to one type of object, i.e. `/zones` returns a list of zones, and `/policies` returns a list of policies, and we don't introduce an "include_policies" or "include_zones" query param. Clients will have to be responsible for querying for the policies associated with the zones as they need for their own records. 

If we do have a need to keep both objects in one endpoint, a few thoughts:

- Send child objects first, then parents, so clients can maintain any foreign key relationships as data comes in. In this case, policies and then zones
- Pagination behaves on the total number of objects, so, e.g. if we have a response with 11 policies and 12 zones, and the request has a limit of 10 objects, the response is in 3 pages: 
    - data: [{policies: [<10 policies>]}, {zones: [<0 zones>]}]
    - data: [{policies: [<1 policy>]}, {zones: [<9 zones>]}]
    - data: [{policies: [<0 policies>]}, {zones: [<3 zones>]}]

**Reply by schnuerle** [reply]
I like your solution for pagination if there are two objects in one endpoint.

**Comment by jiffyclub** [comment]
I agree strongly with @joeb1415's point that endpoints should only return one type of data object. `/zones` returns zones, `/policies` returns policies, and we don't mix them. That has the benefit that we don't need to worry about pagination schemes across multiple response fields and can always handle pagination by concatenating the contents of the single key in the `data` object.

**Reply by schnuerle** [reply]
I'm going to look back and see how we came to adding 2 objects in the /zones endpoint. It may have been for simplicity and getting it all at once, but we may be beyond that with the development of CDS now.

**Reply by schnuerle** [reply]
I've created a new branch and PR for @jiffyclub @zanebclark @joeb1415 @jacobbaskin @brynn to review. 

"Specifically removing the policy objects as an option from the curbs/curb endpoint to align with the JSON API spec, reduce redundancy, improve spec clarity, and allow proper pagination."

- Branch [feature-curbs-policy](https://github.com/openmobilityfoundation/curb-data-specification/tree/feature-curbs-policy/curbs#query-curb-zones)
- PR #67

**Comment by schnuerle** [comment]
Well I can see your point @jiffyclub and @joeb1415, but this configuration has been there for a few months and we are about to launch. We will have to go back and look at the meetings and reasoning about why they are in the same endpoint. I recall some strong feelings about having all policy and curb data in a single endpoint, and not split across two of them. A change may have to wait for the next release.

**Reply by zanebclark** [reply]
Given that both objects returned have an identifier, I don't have a strong opinion on the subject of splitting the two objects out. 

If we support multiple objects in the response, it think it's important that one object be considered "primary" and the non-primary object(s) be included as referenced by the "primary" object. It's acceptable for the non-primary objects to be duplicated across two or more responses. Without designating a "primary" object, it wouldn't be clear to the client how to fetch the next batch. That is, unless limit/offset strategy was employed and the the query was deterministically sorted. I'm under the impression that this isn't an efficient pagination strategy. 

Speaking of pagination issues, how are we going to address new records introduced during the query's paginated execution? This would likely only apply to real-time data.

**Reply by jiffyclub** [reply]
I don't think we're trying to closely follow the JSON API spec, but since we reference it with respect to pagination I think it's a useful reference here and it is pretty clear about the [contents of the `data` key](https://jsonapi.org/format/#document-top-level):

> Primary data MUST be either:
> 
> - a single resource object, a single resource identifier object, or null, for requests that target single resources
> - an array of resource objects, an array of resource identifier objects, or an empty array ([]), for requests that target resource collections

If we were to follow the JSON API spec for the `/zones` endpoint we'd expect the data key to be an array of zones.

[JSON API resource objects](https://jsonapi.org/format/#document-resource-objects) may have a [`relationships` field](https://jsonapi.org/format/#document-resource-object-relationships) that contains information about related resources, which I think would include the policies if this were following that spec. But zones already do include the policy IDs directly within their content, which I think would be considered redundant with the JSON API relationship.

**Reply by schnuerle** [reply]
I think Curbs would be the primary and Policy would be secondary if we keep along this path.

I believe that a solution to new records introduced during the query's paginated execution would be left as an exercise to the implementer. :) MDS has pagination of real time or historic data so we may ask some providers and cities how they handle this. 

Tagging some people to see how they handle latency with MDS when paginating data: @jacobbaskin @jiffyclub @jean-populus @marie-x @avatarneil @quicklywilliam
